### PR TITLE
chore: add `attachCustomSynthesis` private API

### PR DESCRIPTION
**Pull request by rix0rrr**
In "legacy" constructs, constructs had the ability to have themselves
activated during synthesis, so they could add additional output
artifacts to the Cloud Assembly.

In #14974, we made good use of that to have the `PublishAssetAction`
have the ability to emit the buildspec it generated to a file
instead of to the CloudFormation template, improving pipeline
performance.

In constructs v10, and hence CDKv2, that ability got removed since the
Cloud Assembly and "synthesis" are domain-specific concepts that
aren't part of the constructs library.

This now breaks the merge to v2 in #15132, since `PublishAssetAction`
uses a concept that cannot be translated.

------------

Add an AWS CDK-specific mechanism back to have constructs inject
themselves into the synthesis lifecycle, and make `PublishAssetAction`
use that new mechanism.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
Hmmm I obviously need to add unit tests.

Review apart from that, please :)

File: packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts [hunk 1]
```diff
@@ -157,6 +157,10 @@ export class PublishAssetsAction extends CoreConstruct implements codepipeline.I
         CDK_CLI_VERSION: { value: props.cdkCliVersion },
       } : undefined,
     });
+
+    App.addCustomSynthesis(this, {
```
Comment (nija-at) [inline packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts:161]: Why not just use a `Stage.of(this).outdir` and push the build spec file to the cloud assembly in the constructor itself? It feels like we can avoid implementing this in `synthesize()`?

**Review by nija-at — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts:161]: A totally legitimate approach (used by asset bundling for example)

Comment (rix0rrr) [reply packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts:161]: BuildSpec won't be known yet at constructor time.

File: packages/@aws-cdk/core/lib/private/synthesis.ts [hunk 2]
```diff
@@ -41,6 +42,12 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   return builder.buildAssembly();
 }
 
+const CUSTOM_SYNTHESIS_MAP = new WeakMap<constructs.IConstruct, ICustomSynthesis>();
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/synthesis.ts:45]: I suppose we could, but  this feels cleaner. Is there anything specific you're concerned about?

[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/core/lib/private/synthesis.ts:45]: Just healthier design - global state is generally hazardous and should be avoided if possible.

File: packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts:161]: We could use `Lazy`.

File: packages/@aws-cdk/core/lib/app.ts [hunk 3]
```diff
@@ -93,6 +96,18 @@ export class App extends Stage {
...
 
+  /**
+   * Add a custom synthesis for the given construct
+   *
+   * When the construct is being synthesized, this allows it to add additional items
+   * into the Cloud Assembly output.
+   *
+   * This feature is intended for library authors; CDK users should almost never need to use this.
+   */
+  public static addCustomSynthesis(construct: Construct, synthesis: ICustomSynthesis): void {
```
Comment (nija-at) [inline packages/@aws-cdk/core/lib/app.ts:107]: This API design allows 'any' construct to modify the synthesis output of any other construct. Possibility for abuse here by construct libraries and inject side-effect behaviour that's hard to account for.

Maybe we could add a method to `Resource` instead, or a magic method?

Neverthless, my preference is for none of these but instead to generate the buildspec in the constructor in combination with `Lazy`.

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/app.ts:107]: > This API design allows 'any' construct to modify the synthesis output of any other construct. Possibility for abuse here by construct libraries and inject side-effect behaviour that's hard to account for.

I'm not sure how this is different if we do it at `app.synth()` time or in the constructor? 

Fact of the matter is you're running code on your machine that could use `fs.writeFileSync()` or `subprocess.spawn()` to do anything to any file on your machine. That capability is already there; at least if we provide a promoted API for it, we could build in checks and gates.

> Neverthless, my preference is for none of these but instead to generate the buildspec in the constructor in combination with Lazy.

I'm not seeing how that helps. I think you are proposing to do `fs.writeFileSync()` in the constructor? How will a `Lazy` help?

**Review by nija-at — APPROVED** [review]
Interesting choice. Definitely feels better than the previous rev.
I'm still not a fan of this API but I'll meet you half way there.

**Pull request merged**